utilities/flink/flink_compaction_filter.cc

```diff
@@ -0,0 +1,148 @@
+    "State type: %d, TTL: %d ms, useSystemTime: %d, timestamp_offset: %d",
+    key.ToString().c_str(), existing_value.ToString(true).c_str(), value_type,
+    config->state_type_, config->ttl_, config->useSystemTime_, config->timestamp_offset_);
+
+  // too short value to have timestamp at all
+  const bool tooShortValue = existing_value.size() < config->timestamp_offset_ + TIMESTAMP_BYTE_SIZE;
+
+  const StateType state_type = config->state_type_;
+  const bool value_or_merge = value_type == ValueType::kValue || value_type == ValueType::kMergeOperand;
+  const bool value_state = state_type == StateType::Value && value_type == ValueType::kValue;
+  const bool list_entry = state_type == StateType::List && value_or_merge;
+  const bool toDecide = value_state || list_entry;
+  const bool list_iter = list_entry && list_element_iter_;
+
+  Decision decision = Decision::kKeep;
+  if (!tooShortValue && toDecide) {
+    decision = list_iter ?
+            ListDecide(existing_value, config, new_value) :
+            Decide(data, config->ttl_, config->timestamp_offset_, CurrentTimestamp(config->useSystemTime_), logger_);
+  }
+  Debug(logger_.get(), "Decision: %d", decision);
+  return decision;
+}
+
+CompactionFilter::Decision FlinkCompactionFilter::ListDecide(
+        const Slice& existing_value, const Config* config, std::string* new_value) const {
+  const int64_t current_timestamp = CurrentTimestamp(config->useSystemTime_);
+  std::size_t offset = 0;
+  if (offset < existing_value.size()) {
+    Decision decision = Decide(existing_value.data(), config->ttl_, offset + config->timestamp_offset_, current_timestamp, logger_);
+    if (decision != Decision::kKeep) {
+      offset = ListNextOffset(existing_value, offset, config->ttl_, current_timestamp);
+      if (offset >= JAVA_MAX_SIZE) {
+        return Decision::kKeep;
+      }
+    }
+  }
+  if (offset >= existing_value.size()) {
+    return Decision::kRemove;
+  } else if (offset > 0) {
+    SetUnexpiredListValue(existing_value, offset, new_value);
+    return Decision::kChangeValue;
+  }
+  return Decision::kKeep;
+}
+
+std::size_t FlinkCompactionFilter::ListNextOffset(
+        const Slice& existing_value, const std::size_t offset, const int64_t ttl, const int64_t current_timestamp) const {
+  std::size_t new_offset = list_element_iter_->NextUnexpiredOffset(existing_value, ttl, current_timestamp);
+  if (new_offset >= JAVA_MAX_SIZE || new_offset < offset) {
+    Error(logger_.get(), "Wrong next offset in list iterator: %d -> %d",
+          offset, new_offset);
+    new_offset = JAVA_MAX_SIZE;
+  } else {
+    Debug(logger_.get(), "Next unexpired offset: %d -> %d",
+          offset, new_offset);
+  }
+  return new_offset;
+}
+
+void FlinkCompactionFilter::SetUnexpiredListValue(
+        const Slice& existing_value, std::size_t offset, std::string* new_value) const {
+  new_value->clear();
+  Slice new_value_slice = Slice(existing_value.data() + offset, existing_value.size() - offset);
```

Comment:
Why do we need to first create `new_value_slice` instead assigning to the `new_value` directly from `existing_value`'s ptr + offset?

Reply:
logger uses `Slice.ToString`, but I could refactor it to create the `Slice` only if log level is debug.

Reply:
Sure, but `new_value` is also a string already?

Reply:
true, but it is not zero-terminated in general, as I understand, and I considered hex byte representation for debuging provided by `Slice` impl.